# HG changeset patch
# User Jaroslav Hajek <highegg@gmail.com>
# Date 1273555520 -7200
#      Tue May 11 07:25:20 2010 +0200
# Node ID 6fb954475e21432d95044476c8ae28906427d91e
# Parent  9f0a264d2f602862a2c7fb7c84515e0698067175
fix off-by-1 error in idx_vector (bug #29851)

diff --git a/liboctave/ChangeLog b/liboctave/ChangeLog
--- a/liboctave/ChangeLog
+++ b/liboctave/ChangeLog
@@ -1,8 +1,12 @@
+2010-05-11  Jaroslav Hajek  <highegg@gmail.com>
+
+	* idx-vector.cc (idx_vector::unmask): Fix off-by-1 bug. Add tests.
+
 2010-05-06  Jaroslav Hajek  <highegg@gmail.com>
 
 	* Array.cc (Array::delete_elements (const idx_vector&)): Only call
 	resize1 when deleting the last element of a vector.
 
 2010-05-06  Jaroslav Hajek  <highegg@gmail.com>
 
 	* oct-norm.h: Fix include guard.
diff --git a/liboctave/idx-vector.cc b/liboctave/idx-vector.cc
--- a/liboctave/idx-vector.cc
+++ b/liboctave/idx-vector.cc
@@ -1197,17 +1197,17 @@ idx_vector::unmask (void) const
       const bool *data = r->get_data ();
       octave_idx_type ext = r->extent (0), len = r->length (0);
       octave_idx_type *idata = new octave_idx_type [len];
 
       for (octave_idx_type i = 0, j = 0; i < ext; i++)
         if (data[i]) 
           idata[j++] = i;
 
-      ext = len > 0 ? idata[len - 1] : 0;
+      ext = len > 0 ? idata[len - 1] + 1 : 0;
 
       return new idx_vector_rep (idata, len, ext, r->orig_dimensions (),
                                  DIRECT);
     }
   else
     return *this;
 }
 
@@ -1308,8 +1308,15 @@ INSTANTIATE_SCALAR_VECTOR_REP_CONST (dou
 INSTANTIATE_SCALAR_VECTOR_REP_CONST (octave_int8)
 INSTANTIATE_SCALAR_VECTOR_REP_CONST (octave_int16)
 INSTANTIATE_SCALAR_VECTOR_REP_CONST (octave_int32)
 INSTANTIATE_SCALAR_VECTOR_REP_CONST (octave_int64)
 INSTANTIATE_SCALAR_VECTOR_REP_CONST (octave_uint8)
 INSTANTIATE_SCALAR_VECTOR_REP_CONST (octave_uint16)
 INSTANTIATE_SCALAR_VECTOR_REP_CONST (octave_uint32)
 INSTANTIATE_SCALAR_VECTOR_REP_CONST (octave_uint64)
+
+/*
+
+%!error id=Octave:index-out-of-bounds 1(find([1,1] != 0))
+%!assert ((1:3)(find([1,0,1] != 0)), [1,3])
+
+*/
